pkg/controller/nodes/branch/handler.go

```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
 }
 
 func (b *branchHandler) Setup(ctx context.Context, setupContext handler.SetupContext) error {
+	logger.Debugf(ctx, "BranchNode::Setup: nothing to do")
 	return nil
 }
 
-func (b *branchHandler) Handle(ctx context.Context, executionContext handler.NodeExecutionContext) (handler.Transition, error) {
-	panic("implement me")
+func (b *branchHandler) Handle(ctx context.Context, nCtx handler.NodeExecutionContext) (handler.Transition, error) {
+	logger.Debug(ctx, "Starting Branch Node")
+	branchNode := nCtx.Node().GetBranchNode()
+	if branchNode == nil {
+		return handler.DoTransitionToFailed(errors.IllegalStateError, "Invoked branch handler, for a non branch node."), nil
+	}
+
+	w := nCtx.Workflow()
+	nodeInputs, err := nCtx.InputReader().Get(ctx)
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to read input. Error [%s]", err)
+		return handler.DoTransitionToFailed(errors.RuntimeExecutionError, errMsg), nil
+	}
+
+	if nCtx.NodeStatus().GetPhase() == v1alpha1.NodePhaseNotYetStarted {
+		finalNodeId, err := DecideBranch(ctx, w, nCtx.NodeID(), branchNode, nodeInputs)
+		if err != nil {
+			errMsg := fmt.Sprintf("Branch evaluation failed. Error [%s]", err)
+			return handler.DoTransitionToFailed(errors.IllegalStateError, errMsg), nil
+		}
+
+		branchNodeState := handler.BranchNodeState{FinalizedNodeID: finalNodeId, Phase: v1alpha1.BranchNodeSuccess}
+		err = nCtx.NodeStateWriter().PutBranchNode(branchNodeState)
+		if err != nil {
+			logger.Errorf(ctx, "Failed to store TaskNode state, err :%s", err.Error())
+			return handler.UnknownTransition, err
+		}
+
+		var ok bool
+		finalNode, ok := w.GetNode(*finalNodeId)
+		if !ok {
+			errMsg := fmt.Sprintf("Branch downstream finalized node not found. FinalizedNode [%s]", *finalNodeId)
+			logger.Debugf(ctx, errMsg)
+			return handler.DoTransitionToFailed(errors.DownstreamNodeNotFoundError, errMsg), nil
+		}
+		i := nCtx.NodeID()
+		childNodeStatus := w.GetNodeExecutionStatus(finalNode.GetID())
```

Comment:
@kumare this should be moved to branchstate but in next PR. lets go with this one in this PR